Allow Setting of EVM Address by EOA (#1082)
hashgraph/hedera-improvement-proposal: HIP/hip-1082.md

- Every account on the network has 1 or 2 evm addresses. Either only the Hedera num alias (long zero) or the Hedera num
    alias (long zero) and an EVM address alias (standard ECDSA derived EVM address).
- When a user interacts with the EVM the network will utilize the ECDSA derived EVM address alias on an account if
    present. If not it will utilize the long zero.
- Accounts with an ECDSA derived EVM address are fully compatible with expectations of contracts that utilize the
    ecrecover precompile.
- Accounts without an ECDSA derived EVM address alias will not match the expectation of the ecrecover precompile
    calculations. This has effects on certain DApp behaviour.
- Smart contracts can and do cache the evm address of an account and often use it to assign attributes such as access
    and ownership. As a result it was important to maintain a static alias to avoid unintended differences in
    expectation
- Account key rotation has no impact on the evm address utilized when an account interacts with the EVM

<aside>
🚨 **Open Task:** : Add account, address key relationship diagram for clarity
</aside>

## User stories

> 1. As an existing account with an ED key but no EVM address alias, I would like to set an ECDSA derived EVM address

Reviewer: In this user story, wouldn't the user have to update the ED key to the new ECDSA key that corresponds with the EVM address for 'ecrecover' to function? Maybe I'm misunderstanding the phrase 'correct ecrecover functionality', or is the implication that the key is being updated as well

Author: No, the user would just use their ECDSA key in a wallet to sign. So far as the EVM address can be extracted from the signature the ecercover flow would work

Reviewer 2: Do you mean a scenario when an account has a KeyList containing existing ED key and a new ECDSA key is added that is added there and is linked to this account.

How this wallet ECDSA key will be connected to the account?

    alias on my account that will identify my account on the EVM and allow for the correct `ecrecover` functionality.
2. As an existing account with an EC key but no EVM address alias, I would like to set an ECDSA derived EVM address
    alias on my account that will identify my account on the EVM and allow for the correct `ecrecover` functionality.
3. As an existing account with a complex key but no EVM address alias, I would like to set an ECDSA derived EVM address

>     alias on my account that will identify my account on the EVM and allow for the correct `ecrecover` functionality.

Reviewer: Same concern here as user story 1. Maybe we add the caveat to the stories, that the ECDSA signature that corresponds with the EVM address for ecrecover to function would need to be utilized on signatures for EVM of those transactions and not the key associated with the account on the hedera network.

4. As an account with an EVM address alias set after creation, I would like to set an evm address override to my
    previous long zero address for a `ContractCreate` transaction.
5. As an account with an EVM address alias set after creation, I would like to set an evm address override to my

>     previous long zero address for a `ContractCall` transaction.

Reviewer: Is this out of scope / too complex for this HIPs user story?
As an existing account with an ED key but no EVM address alias, I would like to update my key to an ECDSA key and set my EVM address to be derived from the new ECDSA key.

Author: Out of scope, not too complex build would be enabled by this HIP and current network functionality.
Such a user could rotate their keys before or after setting the EVM address.
The could actually do it in one single CryptoUpdate also.
Key rotation is not tied to alias value

Reviewer 3: Is this resolved @ty-swirldslabs @Nana-EC ?

## Backwards Compatibility

This does not break previous functionality but rather adds new scoped support for a change that is possible at account
creation time. However, it may be said that there is a behaviour change here that is important to note. This change
is accepted as the benefits of unblocking existing accounts outweighs the cons. The ability to override the address
also ensure there's a pathway for EOA to set their long zero address if needed.
Finally, it is appropriate for clients to query mirror node for accurate account details. Clients that do this will
avoid not be impacted by this change.

## Security Implications

Smart contracts have their logic self contained in byte code. Therefore, the implications of address changes are
dictated by each smart contract and not the network. A smart contract may cache an EOA address on first interaction and
utilize that to identify them in future. The cached address could dictate balance access or other privileged access.

As such, an EOA that has interacted with a smart contract using their long zero address and then sets an ECDSA derived

> EVM address after will be responsible for any unexpected interactions with the smart contract

Reviewer 4: Maybe add a suggestion that such accounts should first transfer all contract assets to the EVM address, then update the account's EVM address ?

Author: Updated

---
hip: 1082
title: Allow Setting of EVM Address by EOA
author: Nana Essilfie-Conduah <@Nana-EC>
working-group: Richard Bair <@rbair>, Atul Mahamuni <@atul-hedera>, Ty Smith <ty-swirldslabs>,
Serg Metelin <@sergmetelin>, Ali Nikan <@alinik4n>, Greg Scullard <@gregscullard>
requested-by: Hedera community
type: Standards Track
category: Service
needs-council-approval: Yes
status: Draft
created: 11-18-24
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/1070
updated: 12-06-24
replaces: 631
requires: 583
---

## Abstract

> HIP 583 opened the doors for greater account compatibility with EVM flows by utilizing the account alias to encompass

Reviewer 5: ```suggestion
[HIP 583](https://hips.hedera.com/hip/hip-583) opened the doors for greater account compatibility with EVM flows by utilizing the account alias to encompass
```

Author: Fixed

the EVM address. However, given the static nature of the alias and concerns of complexity, existing accounts were not

> allowed to update their alias and set one if it was null. This HIP aims to rectify this by allowing EOAs who have no

Author: Updated

5. As an existing account with a newly set ECDSA derived EVM address, I want my account to conform with `ecrecover`
    expectations as long as I sign the transaction with the ECDSA key that corresponds to the EVM address alias.
    This key may or may not be the key currently on my account.
5. As an account with an EVM address alias set after creation, I would like to set an evm address override to my
    previous long zero address for a `ContractCreate` transaction.
6. As an account with an EVM address alias set after creation, I would like to set an evm address override to my
    previous long zero address for a `ContractCall` transaction.
  
## Specification

The main change to the network is simple - `CryptoUpdate` transactions will support setting a new ECDSA derived
`evm_address` if the current value is unset.

### Protobuf
```protobuf
message CryptoUpdateTransactionBody {
    ...

    /**
     * EOA 20-byte address to create that is derived from the keccak-256 hash of a ECDSA_SECP256K1 primitive key.
     */
    bytes evm_address = 19;
}
```

Additionally, `ContractCreate` and `ContractCall` EVM transactions will need to support the ability to specify the

> a Hedera Num Alias alias EVM address to use on the transaction. This is because on the HAPI side all aliases are

Reviewer 5: "a Hedera Num Alias alias EVM addres"
Is that correct?

Author: fixed

> a Hedera Num Alias alias EVM address to use on the transaction. This is because on the HAPI side all aliases are

Reviewer 2: typo: a

Author: Fixed

---
hip: 1082
title: Allow Setting of EVM Address by EOA
author: Nana Essilfie-Conduah <@Nana-EC>
working-group: Richard Bair <@rbair>, Atul Mahamuni <@atul-hedera>, Ty Smith <ty-swirldslabs>,

> Serg Metelin <@sergmetelin>, Ali Nikan <@alinik4n>, Greg Scullard <@gregscullard>

Reviewer 6: Should stay on one line so github doesn't complain about invalid yaml when rendering.

Author: Updated

requested-by: Hedera community
type: Standards Track
category: Service
needs-council-approval: Yes
status: Draft
created: 11-18-24
discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/1070
updated: 12-16-24
replaces: 631
requires: 583
---

## Abstract

[HIP 583](https://hips.hedera.com/hip/hip-583) opened the doors for greater account compatibility with EVM flows by
utilizing the account alias to encompass
the EVM address. However, given the static nature of the alias and concerns of complexity, existing accounts were not

> allowed to update their alias and set one if it was `null`. This HIP aims to rectify this by allowing EOAs who have no

Reviewer 6: null is a bit of tech speak

```suggestion
allowed to update their alias and set one if it was missing. This HIP aims to rectify this by allowing EOAs who have no
```

Author: Updated

alias set to set it to an ECDSA key derived evm address value they own.

This HIP also replaces the need for [HIP 631: Account Virtual Addresses](https://hips.hedera.com/hip/hip-631) as it
aimed to resolve this issue as well as provide even greater functionality. A form of HIP 631 could be revisited in the

> future but not with the goal of providing address equivalence with the EVM.

Author: I think the context is still valuable. A reader of a HIP should be able to read the Abstract and get a general idea of the sections of the spec, for discussion. 
I adopted the 1 liner to give the answer to"What" but preserved the 2 short paragraphs to allow the idea to be presented.

## Motivation

The EVM address alias introduced in HIP 583 allowed for ECDSA based accounts to interact with the EVM in the same way
other EVM chains do. Particularly, functionality like the `ecrecover` that relies on the relationship between key and
address (pivotal to many DEXs) was unblocked and allowed greater EVM equivalence on the Hedera network.
However, HIP 583 intentionally provided no solution for accounts with no alias regardless of key type (ED, EC or
complex keys). Though these accounts were afforded the hedera num account alias (long zero), this format would not
allow accounts to utilize the `ecrecover` precompile in the expected manner.
This continued incompatibility has remained and inhibited web3 EVM tooling support.

Providing a solution to this gap is a necessity to make it easy for EVM dApps in the web3 community to come to 

> Hedera without needing to change their smart contracts.

Author: Adopted change

## Rationale

[HIP 631: Account Virtual Addresses](https://hips.hedera.com/hip/hip-631) lays out the challenges faced by accounts
without an ECDSA derived EVM address that want to interact with the EVM in a way that complies with the Hedera native
model and the EVM model. However, HIP 631 offered feature rich complexity that may be difficult for DApps, wallets and
users to follow completely.

Instead of the full HIP 631, an intermediate approach i.e. a HIP 631 Lite (this HIP) could unblock users without an
EVM address. It would enable accounts to interact with the EVM in a compliant manner regardless of the public key on
its account.

In a sense the current network already supports the ability to set the evm address, it just requires the logic to be 
done on account creation. Thus by lifting the limit of not being able to set the alias after account creation, whiles
still requiring that the current alias be unset this HIP ushers in the ability for those previously blocked accounts to

> interact with the EVM as expected.

Author: Adopted change

### Foundational Concepts

The following are some clarifying concepts to increase clarity around aliases
- Every account on the network has 1 or 2 evm addresses. Either only the Hedera num alias (long zero) or the Hedera num
    alias (long zero) and an EVM address alias (standard ECDSA derived EVM address).
- When a user interacts with the EVM the network will utilize the ECDSA derived EVM address alias on an account if
    present. If not it will utilize the long zero.
- Accounts with an ECDSA derived EVM address are fully compatible with expectations of contracts that utilize the
    ecrecover precompile.
- Accounts without an ECDSA derived EVM address alias will not match the expectation of the ecrecover precompile
    calculations. This has effects on certain DApp behaviour.
- Smart contracts can and do cache the evm address of an account and often use it to assign attributes such as access
    and ownership. As a result it was important to maintain a static alias to avoid unintended differences in
    expectation

> - Account key rotation has no impact on the evm address utilized when an account interacts with the EVM

Author: Adopted change

</aside>

## User stories

1. As an existing account with an ED key but no EVM address alias, I would like to set an ECDSA derived EVM address
    alias on my account that will identify my account on the EVM.
2. As an existing account with an EC key but no EVM address alias, I would like to set an ECDSA derived EVM address
    alias on my account that will identify my account on the EVM.
3. As an existing account with a complex key but no EVM address alias, I would like to set an ECDSA derived EVM address
    alias on my account that will identify my account on the EVM.
5. As an existing account with a newly set ECDSA derived EVM address, I want my account to conform with `ecrecover`
    expectations as long as I sign the transaction with the ECDSA key that corresponds to the EVM address alias.
    This key may or may not be the key currently on my account.
5. As an account with an EVM address alias set after creation, I would like to set an evm address override to my
    previous long zero address for a `ContractCreate` transaction.
6. As an account with an EVM address alias set after creation, I would like to set an evm address override to my

>     previous long zero address for a `ContractCall` transaction.

Author: Adopted change

>   

Author: Adopted change

## Specification

The main change to the network is simple - `CryptoUpdate` transactions will support setting a new ECDSA derived

> `evm_address` if the current value is unset.

Author: Adopted change

### Protobuf
```protobuf
message CryptoUpdateTransactionBody {
    ...

>     /**
>      * EOA 20-byte address to create that is derived from the keccak-256 hash of a ECDSA_SECP256K1 primitive key.
>      */
>     bytes evm_address = 19;

Reviewer 7: ```suggestion
    /**
     * An EVM-compatible 20-byte address derived the keccak-256 hash of a ECDSA_SECP256K1 primitive key.
     * The transaction MUST be signed by the corresponding private key, and the `alias` on the account MUST
     * be unset prior to this call. Once set, this value cannot be changed.
     */
    bytes alias = 19;
```

Author: Adopted change

}
```

Additionally, `ContractCreate` and `ContractCall` EVM transactions will need to support the ability to specify the
Hedera Num Alias address to use on the transaction. This is because on the HAPI side all aliases are applicable and

> thus the flexibility of the API is important to maintain.

Reviewer 6: I still don't understand why the HIP allows them to set an ECDSA derived EVM address on their existing account but then it doesn't allow them to use this new value on contract calls and instead requires the long zero form.

And even if Hedera Num Alias is required, why would they specify it as long zero bytes instead of just using `AccountID` in shard.realm.num?

Author: If a valid alias is set where one did not exist the new value is indeed used during Contract calls.
The override option is provided to set the older long-zero address in the case that a user would like to be perceived by the EVM as such this to unblock themselves.
This is why the address format is used and not the AccountID format.

For example, say an account with no alias has interacted with a contract that cached its long zero address. If they set the alias with this HIP and again interacted with the contract they would likely see the new address cached.
If for some reason that account wanted to migrate in contract details from the original long zero address to the new ECDSA derived address they could do so by utilizing this override in addition to the contract logic.

Also, accounts that previously had no ECDSA derived EVM address were likely created via the HAPI flow and it may
be important for an EOA to utilize the long zero address even after setting a new ECDSA derived EVM address.
For transactions that utilize the `EthereumTransaction` a compliant ECDSA derived EVM address is provided by default
when using EVM tools.

```protobuf
message ContractCreateTransactionBody {
    ...
    /**
    * The 20-byte EVM address to use for the given transaction. This address must be the Hedera num alias of the account.
    */

>     bytes evm_address_override = 21;

Reviewer 6: This name seems ambiguous to me. It could be the address of the contract or the sender. What about `from` like in `eth_call`? Or other alternatives like `from_address` or `sender`?

Author: Fair point.
`sender` might be cleaner as in implementation this is setting the EOA `msg.sender` observed by the EVM

> Additionally, `ContractCreate` and `ContractCall` EVM transactions will need to support the ability to specify the
> Hedera Num Alias address to use on the transaction. This is because on the HAPI side all aliases are applicable and
> thus the flexibility of the API is important to maintain.
> Also, accounts that previously had no ECDSA derived EVM address were likely created via the HAPI flow and it may
> be important for an EOA to utilize the long zero address even after setting a new ECDSA derived EVM address.
> For transactions that utilize the `EthereumTransaction` a compliant ECDSA derived EVM address is provided by default
> when using EVM tools.
> 
> ```protobuf
> message ContractCreateTransactionBody {
>     ...
>     /**
>     * The 20-byte EVM address to use for the given transaction. This address must be the Hedera num alias of the account.
>     */
>     bytes evm_address_override = 21;
> }
> 
> message ContractCallTransactionBody {
> 	...
>     /**
>     * The 20-byte EVM address to use for the given transaction. This address must be the Hedera num alias of the account.
>     */
>     bytes evm_address_override = 5;
> }
> ```

Reviewer 7: ```suggestion
The `ContractCall` transaction will allow the user to specify an `evm_address_override` to explicitly specify
the use of the long-zero EVM address to represent the account in the transaction.

```protobuf
message ContractCallTransactionBody {
	...
    /**
     * The 20-byte EVM address to use for the given transaction. This address must be the Hedera "long-zero" alias
     *  of the account making this call.
     */
    bytes evm_address_override = 5;
}
```
```

Author: Adopted change

> No change is needed to expose the evm address as it is covered by the `evm_address` property on `TransactionRecord`

Reviewer 7: We have both `evm_address` and `alias` in the `TransactionRecord`. The documentation for `evm_address` is suspicious:

> The new default EVM address of the account created by this transaction. This field is populated only when the EVM address is not specified in the related transaction body

I'm not exactly sure what that means. We should double check on whether we write the evm_address, alias, or both in the transaction record after an update. And we should move this comment into the previous section for `CryptoUpdateTransactionBody`.

### HSCS System Contract API

> The HAS system contract will need to be updated to add support for a `setEVmAddressAlias()` function to expose the
> `CryptoUpdate` HAPI functionality and allow EOAs to call this logic directly from dApps.
> 

Reviewer 7: ```suggestion
The Hedera Account Service (HAS) system contract will need to be updated to add support for a `setEvmAddressAlias()` function to expose the
`CryptoUpdate` HAPI functionality and allow EOAs to call this logic directly from dApps.

```

Author: Adopted change

### SDK

> SDKs will need to update the `AccountUpdateTransaction` to support the `setAlias()` function as it does on
> `AccountCreateTransaction`. This will populate the `evm_address` in the `CryptoUpdateTransactionBody`to be submitted.
> 
> `ContractCreateTransaction` and `ContractExecuteTransaction` will need to support a new `setEVmAddressOverride()`
> function to set the `evm_address_override` property on the respective `ContractCreate` and `ContractCall` transactions. 
> 
> In addition to updated `AccountUpdateTransaction` functionality the SDK will want to add new methods to streamline
> account key specifications by developers with considerations for alias
> 
> - `setKeyWithAlias(ECDSAKey)` - Explicitly calls out the alias set and converts public key to address in the background
> - `setKeyWithAlias(Key, ECDSAKey)` - Allows for setting the account key and a separate key that the evmAddress should
>     be derived from. A user must sign the transaction with both keys for this flow to be successful.
> - `setKeyWithoutAlias(Key)` - Explicitly calls out that the alias is not set. This is useful for wallets that create
>     accounts with default keys and rotate keys later.

Reviewer 7: ```suggestion
SDKs will need to update the `AccountUpdateTransaction` to support the `setAlias()` function as it does on
`AccountCreateTransaction`. This will populate the `evm_address` in the `CryptoUpdateTransactionBody`to be submitted.

`ContractExecuteTransaction` will need to support a new `setEVmAddressOverride()`
function to set the `evm_address_override` property on the `ContractCall` transaction. 

In addition to updated `AccountUpdateTransaction` functionality the SDK will want to add new methods to
streamline account key specifications by developers with considerations for alias:

- `setKeyWithAlias(ECDSAKey)` - Explicitly calls out the alias will be set and converts the public key to an
   address in the background
- `setKeyWithAlias(Key, ECDSAKey)` - Allows for setting the account key and a separate key that the evmAddress
   should be derived from. A user must sign the transaction with both keys for this flow to be successful.
- `setKeyWithoutAlias(Key)` - Explicitly calls out that the alias is not set. This is useful for wallets that create
    accounts with default keys and rotate keys later.
```

### Mirror Node

Mirror Nodes should be sure to support the setting of the evm address value on an account when processing a

> `CryptoUpdate` transaction. As in the case of a `CryptoCreate` the new value can be retrieved from the `evm_address`

Reviewer 6: This doesn't mention changes around processing contract call and create.

Author: The CN should ensure that the override address is adopted fully by the EVM, so there should be no changes to Mirror Node processing on those calls.
The `sender` value in `ContractFunctionResult` should be that of the override utilized in the transaction body.
Do you see any processing changes as a result of this or in other areas?
I'm happy to call them out or address them if they were missed

> `CryptoUpdate` transaction. As in the case of a `CryptoCreate` the new value can be retrieved from the `evm_address`

Reviewer 6: Can we detail any API changes? My assumption that there's no new fields but that existing fields in `/api/v1/accounts/{id}` will show the updated evm_address and the existing `from` in contract result APIs will show the override?

Author: Correct, the `evm_address` for an account will show the new value.
New contract results would also show the override.
Older accounts values and contract results should show the long zero if that's what was in the record file for correctness.
As such there shouldn't be any API field changes.
DO you see otherwise?

Added some details for this

> Mirror Nodes should be sure to support the setting of the evm address value on an account when processing a
> `CryptoUpdate` transaction. As in the case of a `CryptoCreate` the new value can be retrieved from the `evm_address`
> property on `TransactionRecord`.
> If no custom logic already exists to ignore this Mirror Nodes will be unimpaired by the change

Reviewer 7: ```suggestion
Mirror Nodes should be sure to support the setting of the evm address / alias value on an account when processing
a `CryptoUpdate` transaction. As in the case of a `CryptoCreate` the new value can be retrieved from the `evm_address` property on `TransactionRecord`.
If no custom logic already exists to ignore this Mirror Nodes will be unimpaired by the change
```

Author: Adopted change

### Wallets

The HIP provides opportunities for increased user functionality.
- Wallets are encouraged to create accounts with an ECDSA derived address alias
- Wallets are encouraged to support the ability for users to set an ECDSA derived alias in the future. This may require
wallets to support multiple keys for users that may decide to interchange between addresses.

## Backwards Compatibility

> This does not break previous functionality but rather adds new scoped support for a change that is possible at account
> creation time. However, it may be said that there is a behaviour change here that is important to note. This change
> is accepted as the benefits of unblocking existing accounts outweighs the cons. The ability to override the address
> also ensure there's a pathway for EOA to set their long zero address if needed.
> Finally, it is appropriate for clients to query mirror node for accurate account details. Clients that do this will
> avoid not be impacted by this change.

Reviewer 7: ```suggestion
Any account interacting with a smart contract using its long-zero address, that subsequently sets an ECDSA derived alias, MAY need to use the `evm_address_override` when working with that smart contract,
if the smart contract maintained a reference to the account. Alternatively, the user may call the smart
contract using the `evm_address_override` to move assets or data contained within the smart contract
from the long-zero address to their new ECDSA-derived address.
```

Author: Adopted change

## Security Implications

> Smart contracts have their logic self contained in byte code. Therefore, the implications of address changes are
> dictated by each smart contract and not the network. A smart contract may cache an EOA address on first interaction and
> utilize that to identify them in future. The cached address could dictate balance access or other privileged access.
> 
> As such, an EOA that has interacted with a smart contract using their long zero address and then sets an ECDSA derived
> EVM address after will be responsible for any unexpected interactions with the smart contract.
> It is advised that an EOA in this case first transfer all in contract storage assets to the EVM address prior to
> setting their alias.

Reviewer 7: ```suggestion
None.
```

Author: Adopted change

## How to Teach This

> SDK examples, blogs and tutorials on docs.hedera.com

Reviewer 7: ```suggestion
SDK examples, blogs and tutorials on docs.hedera.com. We recommend updating all samples and tutorials to use
ECDSA derived aliases in all cases with the matching ECDSA key being used as the account admin key, as this is
the expected normal flow. We strongly recommend only **advanced** users set the alias on existing accounts,
and only once their wallets support it. Otherwise, we recommend setting an alias on an existing account only be
undertaken by wallets or dapps that pre-create accounts but never use them, until given to a user where the alias
and key are then set.
```

Author: Adopted change

## Reference Implementation

## Rejected Ideas

### Implement HIP 631
[HIP 631: Account Virtual Addresses](https://hips.hedera.com/hip/hip-631) would have added the ability add evm address
and specify any one of them as the desired address for use. However, this didn't unblock existing account and offered
likely complex UX that would overcomplicated the issue at this time. Focusing on setting aliases that were already null
and supporting evm address only for accounts that migrated themselves provides a solution that is focused and less

> complex.

Reviewer 7: Is this true? I thought 631 allowed having multiple aliases, and thus would support going from null to an alias as well. The only problem with 631 was that it was complicated by having multiple aliases, and we decided to try a simpler approach with just 1 alias (in addition to the long-zero).

Author: [HIP 631](https://hips.hedera.com/hip/hip-631) specifically reserved a different space in state for the multiple virtual addresses but maintained a single account.alias field as immutable

> With Virtual Addresses, keys remain the signing permission authority and can be changed but Virtual Addresses (not tied to the alias) provide the value for a transactions EVM address.

Pulled from HIP 631, also the account diagram notes virtual addresses as a different field altogether. We never stated the ability to set a the null alias, so even with multiple virtual addresses an account could still have a null alias

## Open Issues

While a HIP is in draft, ideas can come up which warrant further discussion. Those ideas should be recorded so people
know that they are being thought about but do not have a concrete resolution. This helps make sure all issues required

> for the HIP to be ready for consideration are complete and reduces people duplicating prior discussions.

Reviewer 7: Either we should remove this or list the open issues.

Author: Removed

## References

> 

Reviewer 7: It would be good to add the references to the other HIPs.

Author: Added

Permits accounts without an alias to set one.

[HIP 583](https://hips.hedera.com/hip/hip-583) opened the doors for greater account compatibility with EVM flows by
utilizing the account alias to encompass the EVM address. However, given the static nature of the alias and concerns of
complexity, existing accounts with a missing alias value were not permitted to set their alias.
This HIP aims to rectify this by allowing EOAs who have no alias set to set it to an ECDSA key derived evm address
value they can prove ownership of.

This HIP also replaces the need for [HIP 631: Account Virtual Addresses](https://hips.hedera.com/hip/hip-631) as it
aimed to resolve this issue as well as provide even greater functionality. A form of HIP 631 could be revisited in the
future but not with the goal of providing address equivalence with the EVM.

## Motivation

On Ethereum, every account is addressed by an address which is derived from an ECDSA public key. The
EVM also has a specific precompile function, `ecrecover`, that given an ECDSA signature can extract the public key
and convert that public key into an address. This allows an EVM program to figure out the address of
transaction signers.

> Hiero accounts are not limited to ECDSA keys, and Hiero accounts can rotate their keys. Both of these

Reviewer 8: This section and the following two need some additional blank lines to introduce paragraphing in the rendered version.

referred to simply as the "account key" or "key"), with an EVM compatible "alias". This alias could match
exactly what an address would be on Ethereum. The user could call a smart contract, signed with the
private key associated with their alias, and the smart contract could use `ecrecover` to recover the public
key, derive the alias, and lookup the account.
However, not all accounts on Hedera use ECDSA keys or were created after HIP-583 was defined. This HIP adds support for
specifying an alias on an account, so long as that account does not already have a valid alias and the alias to be set
is not already in use on the network. [HIP-631](https://hips.hedera.com/hip/hip-631) defines a more complex set of
rules that would allow multiple aliases to be defined for any given account. This HIP differs from HIP-631 in
that it supports only a single alias. HIP-631 could be implemented compatibly at a later date if desired.

> Note 1: An ECDSA derived address alias is unique across the network. Only 1 account may utilize a given alias as a
time. Attempts to create an account or update an account to use an already in use alias will be rejected by the network.

> > Note 2: Key rotation and aliases are separate concepts. A valid alias when set is immutable and remains for the

Reviewer 9: If the alias is immutable once set and the user loses their ECDSA private key, not only are the EVM assets associated with that key lost, but that Hedera account loses its ability to be associated with a different EVM address.  The user would have to create a new Hedera account, associate it with a new EVM address, and transfer all Hedera assets from the old account to the new account. 

I think this needs to be documented as an undesirable side effect of an immutable association and a disadvantage reduced target instead of the full dynamism in 631.

Author: Additionally, this wouldn't change in an implementation of HIP 631, as the alias is still immutable and the new virtual addresses would again be tied up to a single account at a time. If they wanted to move the address though it would be a CryptoUpdate vs a CryptoDelta + CryptoCreate as is the case of the existing network pre and post this HIP.
HIP 631 could make the UX of moving an alias around easier but it's not difficult today with successive trnasactions.
Assets in an account would still have to be migrated in both cases

Reviewer 9: I am fine with this HIP going through as is, but think there should be a new HIP to extend capability to allow for alias rotation: 
1. The 0x0000000### alias can be rotated to active alias without additional signing. 
2. The active alias can be rotated to any ECDSA derived alias with signing from that key and verification that the alias is not used by any other account. 

This way there are no inaccessible assets caused by permanent alias associations. 

UX problems should be solved at the level of UX, not here.  Tracking possible keys whose alias can be activated and tracking asset associations to ECDSA keys so the user doesn't throw away a key with active assets associated is a UX responsibility.

Reviewer 9: This conversation is resolvable from my perspective.
